### PR TITLE
Implement RS1041 (Compiler extensions should be implemented in assemblies targeting netstandard2.0)

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 1]
```diff
@@ -586,4 +586,13 @@
   <data name="SemanticModelGetDeclaredSymbolAlwaysReturnsNullForFieldMessage" xml:space="preserve">
     <value>A call to 'SemanticModel.GetDeclaredSymbol({0})' will always return 'null'</value>
   </data>
+  <data name="DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFrameworkMessage" xml:space="preserve">
```
Comment (genlu) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx:589]: >DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFramework [](http://example.com/codeflow?start=14&length=67)

make the names consistent, i.e. `DoNotDeclareCompilerFeatureInAssemblyWithUnsupportedTargetFramework vs `DoNotRegisterCompilerTypesWithBadTargetFramework`

File: src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzerTests.cs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            DiagnosticAnalyzer,
+            DiagnosticSuppressor,
+            ISourceGenerator,
+            IIncrementalGenerator,
+        }
+
+        public enum SupportedTargetFramework
+        {
+            //NetStandard1_0,
+            //NetStandard1_1,
```
Comment (genlu) [inline src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzerTests.cs:43]: why are those commented out?

Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzerTests.cs:43]: I don't have a way to test these.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx (hunk 1 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx:589]: ➡️ Updated

**Review by genlu — APPROVED** [review]

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+                if (appliedTargetFrameworkAttribute.ConstructorArguments.IsEmpty)
+                {
+                    return;
+                }
+
+                string displayName;
+                switch (appliedTargetFrameworkAttribute.ConstructorArguments[0].Value as string)
+                {
+                    case ".NETStandard,Version=v1.0":
```
Comment (jaredpar) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs:65]: The compiler already has the ability to pass MSBuild properties through as analyzer config options. By default we pass none. I'm starting to wonder if we should automatically pass a few ones like `<TargetFramework>` here. There are a number of analyzers that are interested in TFM. Looking at it that way doesn't require an type of semantic model access.

Comment (genlu) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs:65]: Is this done by project-system?

Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs:65]: This repository has the ability to (and does in some cases) add such properties to the list of analyzer-exposed values. I can convert the analyzer to use this if desired.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+                    context =>
+                    {
+                        var namedType = (INamedTypeSymbol)context.Symbol;
+                        if (!IsRegisteredExtension(namedType, diagnosticAnalyzer, diagnosticAnalyzerAttribute, out var applicationSyntaxReference)
+                            && !IsRegisteredExtension(namedType, sourceGeneratorInterface, generatorAttribute, out applicationSyntaxReference)
+                            && !IsRegisteredExtension(namedType, incrementalGeneratorInterface, generatorAttribute, out applicationSyntaxReference))
+                        {
+                            // This is not a compiler extension
+                            return;
+                        }
```
Comment (Youssef1313) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs:98]: Is it intentional to not check `CodeFixProvider`?

Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/CompilerExtensionTargetFrameworkAnalyzer.cs:98]: ➡️ Yes. CodeFixProvider is an IDE feature, not a compiler feature.

**Pull request merged**